vmware/docker-volume-vsphere
# Remove symlink when removing vmgroup without "--remove-volume" option.

Fixed #1452
Found this issue when I work on my automation test. 
Right now, the following tests are done manually. Later another PR for my automation test will cover this test too.

1. create a vmgroup "test_vmgroup_1" and add two VMs to it
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup create --name=test_vmgroup_1 --vm-list=ubuntu-VM0.0,ubuntu-VM1.0 --default-datastore=sharedVmfs-0
vmgroup 'test_vmgroup_1' is created. Do not forget to run 'vmgroup vm add' to add vm to vmgroup.
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup ls
Uuid                                  Name            Description                Default_datastore  VM_list                   
------------------------------------  --------------  -------------------------  -----------------  -------------------------  
11111111-1111-1111-1111-111111111111  _DEFAULT        This is a default vmgroup  _VM_DS                                       
46d0c05d-f164-484c-8976-7420fad35e62  test_vmgroup_1                             sharedVmfs-0       ubuntu-VM0.0,ubuntu-VM1.0  
```

2. create a volume "vmgroup1-vol1" for vmgroup "test_vmgroup1"
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create --driver=vsphere --name=vmgroup1-vol1
vmgroup1-vol1
root@sc-rdops-vm02-dhcp-52-237:~# 
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere:latest      vmgroup1-vol1@sharedVmfs-0
```

3.  check the symlink for vmgroup "test_vmgroup_1" which points to the right path /vmfs/volume/datastore_name/tenant_uuid
```
[root@sc-rdops-vm17-dhcp-3-236:~] ls -al /vmfs/volumes/sharedVmfs-0/dockvols/
total 1048
drwxr-xr-x    1 root     root           560 Jun 21 22:20 .
drwxr-xr-t    1 root     root          3080 Jun 20 23:22 ..
drwxr-xr-x    1 root     root           700 Jun 21 22:20 46d0c05d-f164-484c-8976-7420fad35e62
lrwxrwxrwx    1 root     root            72 Jun 21 22:20 test_vmgroup_1 -> /vmfs/volumes/sharedVmfs-0/dockvols/46d0c05d-f164-484c-8976-7420fad35e62
```

4. remove VMs from vmgroup "test_vmgroup_1" and then remove the vmgroup without "--remove-volume" option
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup vm rm --name=test_vmgroup_1 --vm-list=ubuntu-VM0.0,ubuntu-VM1.0 
vmgroup vm rm succeeded
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup rm --name=test_vmgroup_1
vmgroup rm succeeded
```

5. run ```vmdkops_admin volume ls```, can see this volume, but does not belong to any vmgroup
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume         Datastore     VMGroup  Capacity  Used  Filesystem  Policy  Disk Format  Attached-to  Access      Attach-as               Created By    Created Date              
-------------  ------------  -------  --------  ----  ----------  ------  -----------  -----------  ----------  ----------------------  ------------  ------------------------  
vmgroup1-vol1  sharedVmfs-0  N/A      100MB     13MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM0.0  Wed Jun 21 22:20:05 2017  
```

6. check the symlink, symlink for this vmgroup has been removed
```
[root@sc-rdops-vm17-dhcp-3-236:~] ls -al /vmfs/volumes/sharedVmfs-0/dockvols/
total 1040
drwxr-xr-x    1 root     root           420 Jun 21 22:21 .
drwxr-xr-t    1 root     root          3080 Jun 20 23:22 ..
drwxr-xr-x    1 root     root           700 Jun 21 22:20 46d0c05d-f164-484c-8976-7420fad35e62
```

7. recreate the vmgroup with the same name "test_vmgroup_1"
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup create --name=test_vmgroup_1 --vm-list=ubuntu-VM0.0,ubuntu-VM1.0 --default-datastore=sharedVmfs-0
vmgroup 'test_vmgroup_1' is created. Do not forget to run 'vmgroup vm add' to add vm to vmgroup.
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] 
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vmgroup ls
Uuid                                  Name            Description                Default_datastore  VM_list                   
------------------------------------  --------------  -------------------------  -----------------  -------------------------  
11111111-1111-1111-1111-111111111111  _DEFAULT        This is a default vmgroup  _VM_DS                                       
80bcaf71-6b96-4968-b03c-0b134475a687  test_vmgroup_1                             sharedVmfs-0       ubuntu-VM0.0,ubuntu-VM1.0  
```

8. create a new volume, ```docker volume ls``` can list the right volume
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create --driver=vsphere --name=new-vmgroup1-vol1
new-vmgroup1-vol1
root@sc-rdops-vm02-dhcp-52-237:~# 
root@sc-rdops-vm02-dhcp-52-237:~# 
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere:latest      new-vmgroup1-vol1@sharedVmfs-0
``` 

9. run ```vmdkops_admin volume ls```, and can see both volumes
```
[root@sc-rdops-vm17-dhcp-3-236:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume             Datastore     VMGroup         Capacity  Used  Filesystem  Policy  Disk Format  Attached-to  Access      Attach-as               Created By    Created Date              
-----------------  ------------  --------------  --------  ----  ----------  ------  -----------  -----------  ----------  ----------------------  ------------  ------------------------  
vmgroup1-vol1      sharedVmfs-0  N/A             100MB     13MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM0.0  Wed Jun 21 22:20:05 2017  
new-vmgroup1-vol1  sharedVmfs-0  test_vmgroup_1  100MB     13MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM0.0  Wed Jun 21 22:22:50 2017  
```

10. check the symlink, symlink for this vmgroup link to the path /vmfs/volumes/datastore_name/new_tenant_uuid

```
[root@sc-rdops-vm17-dhcp-3-236:~] ls -al /vmfs/volumes/sharedVmfs-0/dockvols/
total 1056
drwxr-xr-x    1 root     root           700 Jun 21 22:22 .
drwxr-xr-t    1 root     root          3080 Jun 20 23:22 ..
drwxr-xr-x    1 root     root           700 Jun 21 22:20 46d0c05d-f164-484c-8976-7420fad35e62
drwxr-xr-x    1 root     root           700 Jun 21 22:22 80bcaf71-6b96-4968-b03c-0b134475a687
lrwxrwxrwx    1 root     root            72 Jun 21 22:22 test_vmgroup_1 -> /vmfs/volumes/sharedVmfs-0/dockvols/80bcaf71-6b96-4968-b03c-0b134475a687
```
